### PR TITLE
Add circe module

### PR DESCRIPTION
**Pull request by moradology**
This PR adds a very simple circe module which allows reading/writing of `io.circe.Json` instances as segments of a configuration.

Note: The implicit search in `tut` didn't seem to like my `Decoder` without wrapping things up in an object (I assume as a result of REPL limitations). As a result, the second example's code looks a bit strange

### CODE REVIEW

File: modules/circe/src/main/tut/README.md
Comment (jcazevedo) [inline modules/circe/src/main/tut/README.md]: Since `Either` only became right-biased in Scala 2.12.x, we should add `.right` to these in order for this to compile in Scala 2.11.x.

File: modules/circe/src/main/tut/README.md
Comment (jcazevedo) [inline modules/circe/src/main/tut/README.md]: This should mention `ConfigReader` to be consistent with the example below.

File: modules/circe/src/test/scala/pureconfig/circe/CirceSuite.scala
Comment (jcazevedo) [inline modules/circe/src/test/scala/pureconfig/circe/CirceSuite.scala]: We should name this spec accordingly. Also can you move this file to the folder that mimics the package structure (i.e. `/modules/circe/src/test/scala/pureconfig/module/circe/CirceSuite.scala`)?

File: modules/circe/src/test/scala/pureconfig/circe/CirceSuite.scala
Comment (jcazevedo) [inline modules/circe/src/test/scala/pureconfig/circe/CirceSuite.scala]: We shouldn't need to make `Config` go through a `ConfigWriter` of `ConfigValue`. Comparing to `config.root()` should be enough:

```scala
ConfigWriter[JsonConf].to(JsonConf(confJson)) shouldEqual config.root()
```

File: modules/circe/src/main/tut/README.md
Comment (jcazevedo) [inline modules/circe/src/main/tut/README.md]: Is this object really necessary? I seem to be able to make `tut` happy without it.

File: modules/circe/src/main/tut/README.md
Comment (jcazevedo) [inline modules/circe/src/main/tut/README.md]: This sample `Decoder` is slightly convoluted. Also, I don't find the use case of producing a `ConfigReader` from a `Decoder` very compelling. Converting to an intermediate format is inefficient and makes you lose some information relevant for error reporting. I'm not sure if we should mention this in the README of the module.

File: modules/circe/src/main/scala/pureconfig/module/circe/package.scala
Comment (jcazevedo) [inline modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: I prefer relying on `ConfigValue`'s `valueType()` to build a `Json` instead of converting the `ConfigValue` to a string with the JSON representation and going through Circe's parser. It would avoid this returning an `Either` (we should always be able to produce a `Json` from a `ConfigValue`) and would also remove the need for the `circe-parser` dependency in this module.

File: modules/circe/build.sbt
Comment (jcazevedo) [inline modules/circe/build.sbt]: We can use `0.11.1`.

File: modules/circe/src/main/tut/README.md
Comment (jcazevedo) [inline modules/circe/src/main/tut/README.md]: We should update this to `0.11.0` to be aligned with the current version of PureConfig.

File: modules/circe/src/main/scala/pureconfig/module/circe/package.scala
Comment (jcazevedo) [inline modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: This and the `circeJsonWriter` can be `implicit val`s.

**Review by jcazevedo — CHANGES_REQUESTED** [review]
Thanks for your contribution, @moradology! Please take a look at my comments below.

File: modules/circe/src/main/tut/README.md
Comment (moradology) [reply modules/circe/src/main/tut/README.md]: Yeah, the example is not super motivated. The reason I found this useful was that I had a somewhat complex recursive structure and that I'd written a `Decoder` for and didn't really want to rewrite that (performance of serialization/deserialization wasn't a major concern in my use case).

If this example will generally mislead, I'm happy to remove it

File: modules/circe/src/main/tut/README.md
Comment (moradology) [reply modules/circe/src/main/tut/README.md]: It may not be - I could have misinterpreted the compiler's unhappiness with an earlier use of `???` for the `Decoder` implementation

File: modules/circe/src/main/scala/pureconfig/module/circe/package.scala
Comment (moradology) [reply modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: Point of clarification, as I'm having some difficulty seeing how you avoid using the `circe` parser. Let's say I have `LIST` from `ConfigValue.valueType`. Calling `ConfigValue.unwrapped` on such a structure produces a `List[Object]`, where the children `Object`s have *already* been unwrapped. Digging around the API, it wasn't obvious to me how to avoid this (any of the methods I could find that would have allowed me to handle recursion myself were hidden away in private classes)

Is there something I'm overlooking?

Comment (jcazevedo) [reply modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: I think we need to explicitly cast `ConfigValue` to `ConfigList` when the `valueType` is `LIST`. `ConfigList` then lets you handle each individual element as a `ConfigValue`. It should be reasonably safe to do so in those cases, and it's something that we already resorted to in PureConfig's core.

File: modules/circe/src/main/scala/pureconfig/module/circe/package.scala
Comment (moradology) [inline modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: I'm sure there are strong opinions about this one way or the other but I decided to just follow JS behavior here as a first pass

**Comment by moradology** [comment]
I think this is ready to be looked at once more; the only thing I suspect will need some input on is the handling of the `NaN` as `Json.Null` called out above

**Comment by jcazevedo** [comment]
Sorry for the lack of activity in this, @moradology. I'll take another look at the PR during this week.

**Comment by moradology** [comment]
No problem - everyone's busy with their own stuff. I just wanted to bring it to your attention in case you forgot (and anyway, this project has been very welcoming/responsive compared to most)

File: modules/circe/src/main/scala/pureconfig/module/circe/package.scala
Comment (jcazevedo) [reply modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: I'm OK with `Json.Null` here. It is consistent with what the default `Encoder` for `Double` does in circe. However, we also need to take infinity into account. I think we can remove this fallback and use `Json.fromDoubleOrNull` instead.

File: modules/circe/src/test/scala/pureconfig/circe/CirceSuite.scala
Comment (jcazevedo) [inline modules/circe/src/test/scala/pureconfig/circe/CirceSuite.scala]: Can you please move this to `modules/circe/src/test/scala/pureconfig/module/circe/CirceSuite.scala`?

File: modules/circe/src/main/scala/pureconfig/module/circe/package.scala
Comment (jcazevedo) [inline modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: This can probably be made private.

File: modules/circe/src/main/scala/pureconfig/module/circe/package.scala
Comment (jcazevedo) [inline modules/circe/src/main/scala/pureconfig/module/circe/package.scala]: This can probably be made private.

**Review by jcazevedo — CHANGES_REQUESTED** [review]
@moradology: Sorry for the delay in reviewing this. I just left comments on some minor issues below.

**Review by jcazevedo — APPROVED** [review]
@moradology: Thanks for going through the suggested changes! This looks good to me. Let's just wait for another approval from either @ruippeixotog or @leifwickland before merging.

File: modules/circe/src/main/tut/README.md
Comment (ruippeixotog) [inline modules/circe/src/main/tut/README.md]: Nit: can you remove the double space?

File: modules/circe/src/main/tut/README.md
Comment (ruippeixotog) [inline modules/circe/src/main/tut/README.md]: You don't seem to be using this anywhere.

**Review by ruippeixotog — CHANGES_REQUESTED** [review]
LGTM, just a few nitpicks. Thanks for doing this!

**Comment by ruippeixotog** [comment]
One more thing, can you add an entry for pureconfig-circe on [docs/src/main/tut/docs/library-integrations.md](https://github.com/pureconfig/pureconfig/blob/master/docs/src/main/tut/docs/library-integrations.md)?

**Comment by moradology** [comment]
Thanks for having such a strong process for working with open source contributors (and sorry for the delay in getting back to this)

File: modules/circe/README.md
Comment (jcazevedo) [inline modules/circe/README.md]: You also need to fix this in the [tut source](https://github.com/pureconfig/pureconfig/blob/c9bc53861f5581b9b1d917f7d360d7b7ecb04fa5/modules/circe/src/main/tut/README.md).

**Review by jcazevedo — CHANGES_REQUESTED** [review]

File: modules/circe/README.md
Comment (moradology) [reply modules/circe/README.md]: Rebased with this fix

**Review by jcazevedo — APPROVED** [review]

**Review by ruippeixotog — APPROVED** [review]

**Pull request merged**
